File: embedded/fusion/fusion_controller.cpp
#include "fusion_controller.hpp"
#include <iostream>
#include <cmath>
#include <chrono>

namespace noiselessx {
namespace fusion {

FusionController::FusionController(const FusionConfig& config)
    : config_(config),
      current_mode_(FusionMode::NORMAL),
      previous_mode_(FusionMode::NORMAL),
      current_lambda_(0.5f),
      current_envelope_gain_(1.0f),
      previous_envelope_gain_(1.0f)
{
}

void FusionController::reset() {
    current_mode_ = FusionMode::NORMAL;
    previous_mode_ = FusionMode::NORMAL;
    current_lambda_ = 0.5f;
    current_envelope_gain_ = 1.0f;
    previous_envelope_gain_ = 1.0f;
    manual_bypass_ = false;
    manual_error_ = false;
    frame_count_ = 0;
    fault_count_ = 0;
    last_fault_message_.clear();
    last_processing_time_us_ = 0.0;
}

FusionMode FusionController::evaluate_mode(const FusionInput& input) const noexcept {
    // 1. Unrecoverable hardware failure / both audio paths unavailable
    if (manual_error_ || !input.hardware_available || (!input.ai_available && !input.nlms_available)) {
        return FusionMode::ERROR;
    }

    // 2. Operator-triggered bypass or explicit bypass request
    if (manual_bypass_ || input.bypass_requested) {
        return FusionMode::BYPASS;
    }

    // 3. Component faults (high priority before signal quality heuristics)
    // AI inference failure -> DEGRADED (NLMS-only output)
    if (!input.ai_available) {
        return FusionMode::DEGRADED;
    }

    // Reference mic dropout or clock drift failure -> NLMS_FAULT (AI-only output)
    if (!input.nlms_available) {
        return FusionMode::NLMS_FAULT;
    }

    // 4. Signal condition heuristics (both AI and NLMS available)
    // Impulse noise detection takes precedence to guard listener hearing
    if (input.impulse_probability >= config_.impulse_threshold) {
        return FusionMode::IMPULSE;
    }

    // Low AI confidence -> lean more heavily on NLMS
    if (input.ai_confidence < config_.low_confidence_threshold) {
        return FusionMode::LOW_CONFIDENCE;
    }

    return FusionMode::NORMAL;
}

AudioFrame FusionController::fuse(const FusionInput& input) {
    const auto start_time = std::chrono::steady_clock::now();
    ++frame_count_;

    // 1. Explicit State Machine Transition Evaluation
    const FusionMode next_mode = evaluate_mode(input);
    if (next_mode != current_mode_) {
        previous_mode_ = current_mode_;
        current_mode_ = next_mode;

        // Log specific fault transitions
        if (current_mode_ == FusionMode::NLMS_FAULT) {
            log_fault_event(FusionMode::NLMS_FAULT,
                "Reference (error) mic dropped out or drift exceeded tolerance. Falling back to AI-only path.");
        } else if (current_mode_ == FusionMode::DEGRADED) {
            log_fault_event(FusionMode::DEGRADED,
                "AI speech enhancement inference failing. Falling back to NLMS residual-only path.");
        } else if (current_mode_ == FusionMode::ERROR) {
            log_fault_event(FusionMode::ERROR,
                "Audio hardware unavailable or dual microphone paths simultaneously failed.");
        }
    }

    // 2. Compute Target Impulse Gain (always calculated for protection)
    const float raw_prob = std::clamp(input.impulse_probability, 0.0f, 1.0f);
    float target_gain = 1.0f - (config_.impulse_beta * raw_prob);
    target_gain = std::max(target_gain, config_.impulse_gain_floor);

    // 3. Dispatch to explicit state handler
    AudioFrame out_frame;
    switch (current_mode_) {
        case FusionMode::NORMAL:
            out_frame = process_normal(input, target_gain);
            break;
        case FusionMode::LOW_CONFIDENCE:
            out_frame = process_low_confidence(input, target_gain);
            break;
        case FusionMode::IMPULSE:
            out_frame = process_impulse(input, target_gain);
            break;
        case FusionMode::DEGRADED:
            out_frame = process_degraded(input, target_gain);
            break;
        case FusionMode::NLMS_FAULT:
            out_frame = process_nlms_fault(input, target_gain);
            break;
        case FusionMode::BYPASS:
            out_frame = process_bypass(input);
            break;
        case FusionMode::ERROR:
            out_frame = process_error(input);
            break;
    }

    out_frame.frame_index = frame_count_;
    out_frame.timestamp_ns = input.ai_output.timestamp_ns != 0 
        ? input.ai_output.timestamp_ns 
        : static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::nanoseconds>(
            std::chrono::steady_clock::now().time_since_epoch()).count());

    // Record processing latency
    const auto end_time = std::chrono::steady_clock::now();
    last_processing_time_us_ = std::chrono::duration<double, std::micro>(end_time - start_time).count();

    return out_frame;
}

float FusionController::compute_dynamic_lambda(float ai_conf, float vad_prob, bool low_confidence_mode) const noexcept {
    const float conf = std::clamp(ai_conf, 0.0f, 1.0f);
    const float vad = std::clamp(vad_prob, 0.0f, 1.0f);
    
    // Dynamic lambda computation: lambda = clamp(ai_confidence * vad_probability, lambda_min, lambda_max)
    float lambda = std::clamp(conf * vad, config_.lambda_min, config_.lambda_max);

    if (low_confidence_mode) {
        // In LOW_CONFIDENCE mode, scale lambda down to lean heavily on NLMS residual output
        lambda = std::clamp(lambda * config_.low_confidence_scale, config_.lambda_min, 
                            config_.lambda_min + (config_.lambda_max - config_.lambda_min) * 0.25f);
    }

    return lambda;
}

void FusionController::update_gain_envelope(float target_gain) {
    previous_envelope_gain_ = current_envelope_gain_;

    if (target_gain < current_envelope_gain_) {
        // Fast attack phase: transient onset
        current_envelope_gain_ = (config_.attack_alpha * target_gain) + 
                                 ((1.0f - config_.attack_alpha) * current_envelope_gain_);
    } else {
        // Slower recovery / release phase: transient decaying
        current_envelope_gain_ = (config_.release_alpha * target_gain) + 
                                 ((1.0f - config_.release_alpha) * current_envelope_gain_);
    }

    // Guarantee configurable gain floor > 0: audio never hard mutes
    current_envelope_gain_ = std::clamp(current_envelope_gain_, config_.impulse_gain_floor, 1.0f);
}

void FusionController::apply_gain_envelope(AudioFrame& frame, float prev_gain, float curr_gain) {
    const size_t n_samples = frame.size();
    if (n_samples == 0) return;

    if (n_samples == 1) {
        frame[0] *= curr_gain;
        return;
    }

    // Linear sample-accurate interpolation across frame to avoid boundary clicks
    const float step = (curr_gain - prev_gain) / static_cast<float>(n_samples - 1);
    float gain = prev_gain;
    for (size_t n = 0; n < n_samples; ++n) {
        frame[n] *= gain;
        gain += step;
    }
}

AudioFrame FusionController::process_normal(const FusionInput& input, float target_gain) {
    const size_t n_samples = std::max(input.ai_output.size(), input.nlms_output.size());
    AudioFrame out(n_samples, 0.0f);

    // Compute dynamic lambda
    current_lambda_ = compute_dynamic_lambda(input.ai_confidence, input.vad_probability, false);

    // Fusion rule: s_hat[n] = lambda * s_AI[n] + (1 - lambda) * s_NLMS[n]
    for (size_t n = 0; n < n_samples; ++n) {
        const float s_ai = (n < input.ai_output.size()) ? input.ai_output[n] : 0.0f;
        const float s_nlms = (n < input.nlms_output.size()) ? input.nlms_output[n] : 0.0f;
        out[n] = (current_lambda_ * s_ai) + ((1.0f - current_lambda_) * s_nlms);
    }

    // Impulse protection gain envelope
    update_gain_envelope(target_gain);
    apply_gain_envelope(out, previous_envelope_gain_, current_envelope_gain_);

    return out;
}

AudioFrame FusionController::process_low_confidence(const FusionInput& input, float target_gain) {
    const size_t n_samples = std::max(input.ai_output.size(), input.nlms_output.size());
    AudioFrame out(n_samples, 0.0f);

    // Lean more heavily on NLMS output (lower lambda)
    current_lambda_ = compute_dynamic_lambda(input.ai_confidence, input.vad_probability, true);

    for (size_t n = 0; n < n_samples; ++n) {
        const float s_ai = (n < input.ai_output.size()) ? input.ai_output[n] : 0.0f;
        const float s_nlms = (n < input.nlms_output.size()) ? input.nlms_output[n] : 0.0f;
        out[n] = (current_lambda_ * s_ai) + ((1.0f - current_lambda_) * s_nlms);
    }

    update_gain_envelope(target_gain);
    apply_gain_envelope(out, previous_envelope_gain_, current_envelope_gain_);

    return out;
}

AudioFrame FusionController::process_impulse(const FusionInput& input, float target_gain) {
    const size_t n_samples = std::max(input.ai_output.size(), input.nlms_output.size());
    AudioFrame out(n_samples, 0.0f);

    current_lambda_ = compute_dynamic_lambda(input.ai_confidence, input.vad_probability, false);

    for (size_t n = 0; n < n_samples; ++n) {
        const float s_ai = (n < input.ai_output.size()) ? input.ai_output[n] : 0.0f;
        const float s_nlms = (n < input.nlms_output.size()) ? input.nlms_output[n] : 0.0f;
        if (input.nlms_available) {
            out[n] = (current_lambda_ * s_ai) + ((1.0f - current_lambda_) * s_nlms);
        } else {
            out[n] = s_ai;
        }
    }

    // Apply full impulse gain envelope
    update_gain_envelope(target_gain);
    apply_gain_envelope(out, previous_envelope_gain_, current_envelope_gain_);

    return out;
}

AudioFrame FusionController::process_degraded(const FusionInput& input, float target_gain) {
    // AI inference failing -> output NLMS-only (never silence, since NLMS is active)
    current_lambda_ = 0.0f;
    const size_t n_samples = !input.nlms_output.empty() ? input.nlms_output.size() : config_.default_frame_size;
    AudioFrame out(n_samples, 0.0f);

    for (size_t n = 0; n < n_samples; ++n) {
        out[n] = (n < input.nlms_output.size()) ? input.nlms_output[n] : 0.0f;
    }

    // Impulse protection remains active
    update_gain_envelope(target_gain);
    apply_gain_envelope(out, previous_envelope_gain_, current_envelope_gain_);

    return out;
}

AudioFrame FusionController::process_nlms_fault(const FusionInput& input, float target_gain) {
    // Reference mic dropped out / drift error -> fall back to AI-only
    current_lambda_ = 1.0f;
    const size_t n_samples = !input.ai_output.empty() ? input.ai_output.size() : config_.default_frame_size;
    AudioFrame out(n_samples, 0.0f);

    for (size_t n = 0; n < n_samples; ++n) {
        out[n] = (n < input.ai_output.size()) ? input.ai_output[n] : 0.0f;
    }

    // Impulse protection remains active on AI fallback path
    update_gain_envelope(target_gain);
    apply_gain_envelope(out, previous_envelope_gain_, current_envelope_gain_);

    return out;
}

AudioFrame FusionController::process_bypass(const FusionInput& input) {
    // Input routed directly to output with zero modification / trivial delay
    if (!input.raw_input.empty()) {
        return input.raw_input;
    }
    if (!input.nlms_output.empty()) {
        return input.nlms_output;
    }
    if (!input.ai_output.empty()) {
        return input.ai_output;
    }
    return AudioFrame(config_.default_frame_size, 0.0f);
}

AudioFrame FusionController::process_error(const FusionInput& input) {
    // Both hardware paths unavailable: clear error reported, produce silent buffer
    const size_t n_samples = !input.ai_output.empty() ? input.ai_output.size() 
                           : (!input.nlms_output.empty() ? input.nlms_output.size() : config_.default_frame_size);
    return AudioFrame(n_samples, 0.0f);
}

void FusionController::log_fault_event(FusionMode mode, const std::string& reason) {
    ++fault_count_;
    last_fault_message_ = reason;

    if (config_.log_faults) {
        std::cerr << "[FusionController][FAULT] Frame #" << frame_count_ 
                  << " Transition to " << to_string(mode) 
                  << " - Reason: " << reason << std::endl;
    }

    if (fault_logger_) {
        fault_logger_(mode, reason, frame_count_);
    }
}

} // namespace fusion
} // namespace noiselessx

File: ai/fusion/fusion_controller.py
#!/usr/bin/env python3
"""
Dual-Microphone Fusion Controller with Explicit State Machine.
Fuses AI Speech Enhancement output with NLMS Adaptive Cancellation output.
"""

from dataclasses import dataclass, field
from enum import Enum
import time
from typing import Optional, Callable, List
import numpy as np


class FusionMode(Enum):
    NORMAL = "NORMAL"
    LOW_CONFIDENCE = "LOW_CONFIDENCE"
    IMPULSE = "IMPULSE"
    DEGRADED = "DEGRADED"
    NLMS_FAULT = "NLMS_FAULT"
    BYPASS = "BYPASS"
    ERROR = "ERROR"


@dataclass
class AudioFrame:
    samples: np.ndarray = field(default_factory=lambda: np.zeros(0, dtype=np.float32))
    frame_index: int = 0
    timestamp_ns: int = 0

    def __post_init__(self):
        if not isinstance(self.samples, np.ndarray):
            self.samples = np.asarray(self.samples, dtype=np.float32)
        elif self.samples.dtype != np.float32:
            self.samples = self.samples.astype(np.float32)

    def __len__(self) -> int:
        return len(self.samples)

    def __getitem__(self, idx):
        return self.samples[idx]

    def __setitem__(self, idx, val):
        self.samples[idx] = val


@dataclass
class FusionConfig:
    sample_rate: float = 16000.0
    default_frame_size: int = 80
    lambda_min: float = 0.10
    lambda_max: float = 0.95
    low_confidence_threshold: float = 0.45
    low_confidence_scale: float = 0.25
    impulse_threshold: float = 0.60
    impulse_beta: float = 0.90
    impulse_gain_floor: float = 0.10
    attack_alpha: float = 0.85
    release_alpha: float = 0.12
    log_faults: bool = True


@dataclass
class FusionInput:
    ai_output: AudioFrame
    nlms_output: AudioFrame
    ai_confidence: float = 1.0
    impulse_probability: float = 0.0
    vad_probability: float = 0.0
    nlms_available: bool = True
    raw_input: Optional[AudioFrame] = None
    ai_available: bool = True
    hardware_available: bool = True
    bypass_requested: bool = False


class FusionController:
    """
    Dual-Mic Real-time Speech Enhancement & Noise Cancellation Fusion Engine.
    """

    def __init__(self, config: Optional[FusionConfig] = None):
        self.config = config or FusionConfig()
        self.current_mode = FusionMode.NORMAL
        self.previous_mode = FusionMode.NORMAL
        self.current_lambda = 0.5
        self.current_envelope_gain = 1.0
        self.previous_envelope_gain = 1.0
        self.manual_bypass = False
        self.manual_error = False
        self.frame_count = 0
        self.fault_count = 0
        self.last_fault_message = ""
        self.last_processing_time_us = 0.0
        self.fault_logger: Optional[Callable[[FusionMode, str, int], None]] = None

    def reset(self):
        self.current_mode = FusionMode.NORMAL
        self.previous_mode = FusionMode.NORMAL
        self.current_lambda = 0.5
        self.current_envelope_gain = 1.0
        self.previous_envelope_gain = 1.0
        self.manual_bypass = False
        self.manual_error = False
        self.frame_count = 0
        self.fault_count = 0
        self.last_fault_message = ""
        self.last_processing_time_us = 0.0

    def set_bypass(self, bypass: bool):
        self.manual_bypass = bypass

    @property
    def bypass(self) -> bool:
        return self.manual_bypass

    @bypass.setter
    def bypass(self, value: bool):
        self.manual_bypass = value

    def set_manual_error(self, error: bool):
        self.manual_error = error

    def evaluate_mode(self, inp: FusionInput) -> FusionMode:
        # 1. Unrecoverable hardware failure / both paths down
        if self.manual_error or not inp.hardware_available or (not inp.ai_available and not inp.nlms_available):
            return FusionMode.ERROR

        # 2. Operator-triggered or unrecoverable bypass
        if self.manual_bypass or inp.bypass_requested:
            return FusionMode.BYPASS

        # 3. Component faults
        if not inp.ai_available:
            return FusionMode.DEGRADED

        if not inp.nlms_available:
            return FusionMode.NLMS_FAULT

        # 4. Signal condition heuristics
        if inp.impulse_probability >= self.config.impulse_threshold:
            return FusionMode.IMPULSE

        if inp.ai_confidence < self.config.low_confidence_threshold:
            return FusionMode.LOW_CONFIDENCE

        return FusionMode.NORMAL

    def fuse(self, inp: FusionInput) -> AudioFrame:
        t0 = time.perf_counter_ns()
        self.frame_count += 1

        next_mode = self.evaluate_mode(inp)
        if next_mode != self.current_mode:
            self.previous_mode = self.current_mode
            self.current_mode = next_mode

            if self.current_mode == FusionMode.NLMS_FAULT:
                self._log_fault(FusionMode.NLMS_FAULT, "Reference (error) mic dropped out or drift exceeded tolerance.")
            elif self.current_mode == FusionMode.DEGRADED:
                self._log_fault(FusionMode.DEGRADED, "AI speech enhancement inference failing.")
            elif self.current_mode == FusionMode.ERROR:
                self._log_fault(FusionMode.ERROR, "Audio hardware or dual-mic pipeline completely unavailable.")

        # Compute target impulse attenuation gain: g_I = 1 - beta * impulse_probability
        p_impulse = float(np.clip(inp.impulse_probability, 0.0, 1.0))
        target_gain = max(1.0 - (self.config.impulse_beta * p_impulse), self.config.impulse_gain_floor)

        # Dispatch through state machine
        if self.current_mode == FusionMode.NORMAL:
            out_samples = self._process_normal(inp, target_gain)
        elif self.current_mode == FusionMode.LOW_CONFIDENCE:
            out_samples = self._process_low_confidence(inp, target_gain)
        elif self.current_mode == FusionMode.IMPULSE:
            out_samples = self._process_impulse(inp, target_gain)
        elif self.current_mode == FusionMode.DEGRADED:
            out_samples = self._process_degraded(inp, target_gain)
        elif self.current_mode == FusionMode.NLMS_FAULT:
            out_samples = self._process_nlms_fault(inp, target_gain)
        elif self.current_mode == FusionMode.BYPASS:
            out_samples = self._process_bypass(inp)
        elif self.current_mode == FusionMode.ERROR:
            out_samples = self._process_error(inp)
        else:
            out_samples = self._process_normal(inp, target_gain)

        t1 = time.perf_counter_ns()
        self.last_processing_time_us = (t1 - t0) / 1000.0

        ts = inp.ai_output.timestamp_ns if (inp.ai_output and inp.ai_output.timestamp_ns != 0) else t1
        return AudioFrame(samples=out_samples, frame_index=self.frame_count, timestamp_ns=ts)

    def _compute_dynamic_lambda(self, ai_conf: float, vad_prob: float, low_conf: bool = False) -> float:
        conf = float(np.clip(ai_conf, 0.0, 1.0))
        vad = float(np.clip(vad_prob, 0.0, 1.0))
        l_min, l_max = self.config.lambda_min, self.config.lambda_max
        lam = float(np.clip(conf * vad, l_min, l_max))

        if low_conf:
            lam = float(np.clip(lam * self.config.low_confidence_scale, l_min, l_min + (l_max - l_min) * 0.25))

        return lam

    def _update_gain_envelope(self, target_gain: float):
        self.previous_envelope_gain = self.current_envelope_gain
        if target_gain < self.current_envelope_gain:
            # Fast attack
            self.current_envelope_gain = (self.config.attack_alpha * target_gain) + (
                (1.0 - self.config.attack_alpha) * self.current_envelope_gain
            )
        else:
            # Slower release
            self.current_envelope_gain = (self.config.release_alpha * target_gain) + (
                (1.0 - self.config.release_alpha) * self.current_envelope_gain
            )
        self.current_envelope_gain = float(np.clip(self.current_envelope_gain, self.config.impulse_gain_floor, 1.0))

    def _apply_gain_envelope(self, samples: np.ndarray) -> np.ndarray:
        n = len(samples)
        if n == 0:
            return samples
        if n == 1:
            return samples * self.current_envelope_gain
        envelope = np.linspace(self.previous_envelope_gain, self.current_envelope_gain, n, dtype=np.float32)
        return samples * envelope

    def _get_frame_size(self, inp: FusionInput) -> int:
        s_ai = len(inp.ai_output.samples) if inp.ai_output is not None else 0
        s_nlms = len(inp.nlms_output.samples) if inp.nlms_output is not None else 0
        m = max(s_ai, s_nlms)
        return m if m > 0 else self.config.default_frame_size

    def _process_normal(self, inp: FusionInput, target_gain: float) -> np.ndarray:
        n = self._get_frame_size(inp)
        ai = inp.ai_output.samples if len(inp.ai_output.samples) == n else np.resize(inp.ai_output.samples, n)
        nlms = inp.nlms_output.samples if len(inp.nlms_output.samples) == n else np.resize(inp.nlms_output.samples, n)

        self.current_lambda = self._compute_dynamic_lambda(inp.ai_confidence, inp.vad_probability, False)
        fused = (self.current_lambda * ai) + ((1.0 - self.current_lambda) * nlms)

        self._update_gain_envelope(target_gain)
        return self._apply_gain_envelope(fused)

    def _process_low_confidence(self, inp: FusionInput, target_gain: float) -> np.ndarray:
        n = self._get_frame_size(inp)
        ai = inp.ai_output.samples if len(inp.ai_output.samples) == n else np.resize(inp.ai_output.samples, n)
        nlms = inp.nlms_output.samples if len(inp.nlms_output.samples) == n else np.resize(inp.nlms_output.samples, n)

        self.current_lambda = self._compute_dynamic_lambda(inp.ai_confidence, inp.vad_probability, True)
        fused = (self.current_lambda * ai) + ((1.0 - self.current_lambda) * nlms)

        self._update_gain_envelope(target_gain)
        return self._apply_gain_envelope(fused)

    def _process_impulse(self, inp: FusionInput, target_gain: float) -> np.ndarray:
        n = self._get_frame_size(inp)
        ai = inp.ai_output.samples if len(inp.ai_output.samples) == n else np.resize(inp.ai_output.samples, n)
        nlms = inp.nlms_output.samples if len(inp.nlms_output.samples) == n else np.resize(inp.nlms_output.samples, n)

        self.current_lambda = self._compute_dynamic_lambda(inp.ai_confidence, inp.vad_probability, False)
        if inp.nlms_available:
            fused = (self.current_lambda * ai) + ((1.0 - self.current_lambda) * nlms)
        else:
            fused = ai.copy()

        self._update_gain_envelope(target_gain)
        return self._apply_gain_envelope(fused)

    def _process_degraded(self, inp: FusionInput, target_gain: float) -> np.ndarray:
        # AI inference failing -> output NLMS-only (never silence)
        self.current_lambda = 0.0
        n = len(inp.nlms_output.samples) if len(inp.nlms_output.samples) > 0 else self.config.default_frame_size
        nlms = inp.nlms_output.samples.copy() if len(inp.nlms_output.samples) == n else np.zeros(n, dtype=np.float32)

        self._update_gain_envelope(target_gain)
        return self._apply_gain_envelope(nlms)

    def _process_nlms_fault(self, inp: FusionInput, target_gain: float) -> np.ndarray:
        # Reference mic dropped out / drift error -> fall back to AI-only
        self.current_lambda = 1.0
        n = len(inp.ai_output.samples) if len(inp.ai_output.samples) > 0 else self.config.default_frame_size
        ai = inp.ai_output.samples.copy() if len(inp.ai_output.samples) == n else np.zeros(n, dtype=np.float32)

        self._update_gain_envelope(target_gain)
        return self._apply_gain_envelope(ai)

    def _process_bypass(self, inp: FusionInput) -> np.ndarray:
        # Input routed directly to output with zero modification
        if inp.raw_input is not None and len(inp.raw_input.samples) > 0:
            return inp.raw_input.samples.copy()
        if inp.nlms_output is not None and len(inp.nlms_output.samples) > 0:
            return inp.nlms_output.samples.copy()
        if inp.ai_output is not None and len(inp.ai_output.samples) > 0:
            return inp.ai_output.samples.copy()
        return np.zeros(self.config.default_frame_size, dtype=np.float32)

    def _process_error(self, inp: FusionInput) -> np.ndarray:
        n = self._get_frame_size(inp)
        return np.zeros(n, dtype=np.float32)

    def _log_fault(self, mode: FusionMode, reason: str):
        self.fault_count += 1
        self.last_fault_message = reason
        if self.fault_logger:
            self.fault_logger(mode, reason, self.frame_count)
